File: src/ros2_bridge/mappers/foxglove_msgs/color.rs
//! Typed mapper for `foxglove_msgs/msg/Color`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn color_to_bus(msg: ros_env::foxglove_msgs::msg::Color) -> crate::foxglove_msgs::msg::v1::Color {
    crate::foxglove_msgs::msg::v1::Color {
        r: msg.r,
        g: msg.g,
        b: msg.b,
        a: msg.a,
    }
}

pub(crate) fn color_to_ros(bus: crate::foxglove_msgs::msg::v1::Color) -> ros_env::foxglove_msgs::msg::Color {
    ros_env::foxglove_msgs::msg::Color {
        r: bus.r,
        g: bus.g,
        b: bus.b,
        a: bus.a,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsColorMapper;

impl TypedTopicMapper for FoxgloveMsgsColorMapper {
    type Ros = ros_env::foxglove_msgs::msg::Color;
    type Bus = crate::foxglove_msgs::msg::v1::Color;

    fn type_name(&self) -> &'static str {
        "foxglove_msgs/msg/Color"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(color_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(color_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/color.py
"""Generated mapper for `foxglove_msgs/msg/Color`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.foxglove_msgs.msg.v1 import Color as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def color_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.r = msg.r
    bus.g = msg.g
    bus.b = msg.b
    bus.a = msg.a
    return bus


def color_to_ros(bus):
    from foxglove_msgs.msg import Color as RosMsg

    out = RosMsg()
    out.r = bus.r
    out.g = bus.g
    out.b = bus.b
    out.a = bus.a
    return out


class FoxgloveMsgsColorMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from foxglove_msgs.msg import Color as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return color_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return color_to_ros(bus)
